### PR TITLE
`paddle.nn.func.cosine_embedding_loss`

### LINKED ISSUE #16727: cosine_embedding_loss

### PR DESCRIPTION
**Pull request by mobley-trent**
Closes #16727

### CODE REVIEW

**Comment by mobley-trent** [comment]
@hmahmood24 please review this PR

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @mobley-trent 

Sorry for being late to review this PR🙂, somehow didn't notice it.
Would you update  with the latest master branch changes for me to be able to run tests locally.

Thanks😊

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 1]
```diff
@@ -43,3 +45,49 @@ def binary_cross_entropy_with_logits(
...
+        )
+
+    if len(input1.shape) > 2:
+        raise ValueError(
+            "1D target tensor expects 1D or 2D input tensors, but found inputs with"
+            " different sizes"
+        )
+
+    prod_sum = (input1 * input2).sum(axis=-1)
+    mag_square1 = ivy.square(input1).sum(axis=-1) + 10e-12
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:72]: For proper notation we can use
```suggestion
    mag_square1 = ivy.square(input1).sum(axis=-1) + 1e-12
```

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 2]
```diff
@@ -43,3 +45,49 @@ def binary_cross_entropy_with_logits(
...
+    mag_square2 = ivy.square(input2).sum(axis=-1) + 10e-12
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:73]:
```suggestion
    mag_square2 = ivy.square(input2).sum(axis=-1) + 1e-12
```

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 3]
```diff
@@ -43,3 +45,49 @@ def binary_cross_entropy_with_logits(
...
+    cos = prod_sum / denom
+    zeros = ivy.zeros_like(cos)
+    pos = 1 - cos
+    neg = ivy.clip(cos - margin, 0, 0)
+    out_pos = ivy.where(label == 1, pos, zeros)
+    out_neg = ivy.where(label == -1, neg, zeros)
+    out = out_pos + out_neg
+
+    if reduction == "none":
+        out = ivy.expand_dims(out, axis=-1) if out.ndim == 0 else out
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:84]: The expansion isn't necessary as for some cases we need to return `0-D` Tensor if passed `1-D` Tensor

```python3
x = paddle.to_tensor([-1, -1], dtype='float32')
y = paddle.to_tensor([-0.5, -0.5], dtype='float32')
margin = 0.0
reduce = 'mean'
print(paddle.nn.functional.cosine_embedding_loss(x, x, y, margin=margin, reduction=reduce)) # Tensor(shape=[], dtype=float32, place=Place(cpu), stop_gradient=True, 0.)
```

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 4]
```diff
@@ -43,3 +45,49 @@ def binary_cross_entropy_with_logits(
...
+        return out
+    if reduction == "mean":
+        out = ivy.mean(out)
+        out = ivy.expand_dims(out, axis=-1) if out.ndim == 0 else out
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:88]: This will be omitted as well

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 5]
```diff
@@ -43,3 +45,49 @@ def binary_cross_entropy_with_logits(
...
+        return out
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:89]: We can add this return at the end of all if statements to improve the readability

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 6]
```diff
@@ -43,3 +45,49 @@ def binary_cross_entropy_with_logits(
...
+    elif reduction == "sum":
+        out = ivy.sum(out)
+        out = ivy.expand_dims(out, axis=-1) if out.ndim == 0 else out
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:92]: This will also, be omitted

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 7]
```diff
@@ -43,3 +45,49 @@ def binary_cross_entropy_with_logits(
         ret = ivy.multiply(weight, ret)
     ret = reduction(ret).astype(label.dtype)
     return paddle.to_tensor(ret.reshape([-1]))
+
+
+@handle_exceptions
+@to_ivy_arrays_and_back
+@with_supported_dtypes({"2.4.2 and below": ("float32", "float64")}, "paddle")
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:52]: Now we are working with the latest paddle version `2.5.0`

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @mobley-trent 

Below are comments  for changes, also is it possible to check the loss for large dimensions like 5 😅, try to explore if that's possible.
Thanks😊

File: ivy/functional/frontends/paddle/nn/functional/loss.py (hunk 1 above)
Comment (mobley-trent) [reply ivy/functional/frontends/paddle/nn/functional/loss.py:72]: Won't this intefere with the final values ?

**Comment by mobley-trent** [comment]
> Hey @mobley-trent
> 
> Below are comments for changes, also is it possible to check the loss for large dimensions like 5 😅, try to explore if that's possible. Thanks😊

The function only works for 1D or 2D tensors 😅 @hirwa-nshuti

File: ivy/functional/frontends/paddle/nn/functional/loss.py (hunk 1 above)
Comment (hirwa-nshuti) [reply ivy/functional/frontends/paddle/nn/functional/loss.py]: It won’t interfere as this is epsilon value and the notation should be scientific😄

**Review by hirwa-nshuti — APPROVED** [review]
Hey @mobley-trent 

Now the tests are passing and all looks great😊
Feel free to merge this item🙂
